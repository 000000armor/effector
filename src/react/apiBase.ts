import {
  Store,
  is,
  createNode,
  step,
  clearNode,
  scopeBind,
  Scope,
  Unit,
  Event,
  Node,
} from 'effector'
import React from 'react'
import {useSyncExternalStore} from 'use-sync-external-store/shim'
import {useSyncExternalStoreWithSelector} from 'use-sync-external-store/shim/with-selector'
import {throwError} from './throw'
import {createWatch} from './createWatch'
import {withDisplayName} from './withDisplayName'

const stateReader = <T>(store: Store<T>, scope?: Scope) =>
  scope ? scope.getState(store) : store.getState()
const basicUpdateFilter = <T>(upd: T, oldValue: T) => upd !== oldValue
const keysEqual = (a?: readonly any[], b?: readonly any[]) => {
  if (!a || !b || a.length !== b.length) return false

  let isEqual = true

  for (let i = 0; i < a.length; i++) {
    if (a[i] !== b[i]) {
      isEqual = false
      break
    }
  }

  return isEqual
}

export function useStoreBase<State>(store: Store<State>, scope?: Scope) {
  if (!is.store(store)) throwError('expect useStore argument to be a store')

  const subscribe = React.useCallback(
    (cb: () => void) => createWatch(store, cb, scope),
    [store, scope],
  )
  const read = React.useCallback(
    () => stateReader(store, scope),
    [store, scope],
  )
  const currentValue = useSyncExternalStore(subscribe, read, read)

  return currentValue
}

export function useUnitBase<Shape extends {[key: string]: Unit<any>}>(
  shape: Shape,
  scope?: Scope,
) {
  const isShape = !is.unit(shape) && typeof shape === 'object'
  const normShape = isShape ? shape : {unit: shape}
  const isList = Array.isArray(normShape)
  const entries = Object.entries(normShape)
  const [current, storeMap, stores] = React.useMemo(() => {
    const initial: Record<string, any> = isList ? [] : {}
    const stores: Store<any>[] = []
    const storeMap: Record<string, string> = {}
    for (const [key, value] of entries) {
      if (is.store(value)) {
        if (stores.includes(value)) {
          throwError(`useUnit store at key "${key}" is already exists in shape`)
        }
        stores.push(value)
        /** note that this allows only one occurence of the store */
        storeMap[(value as any).graphite.id] = key
        initial[key] = stateReader(value, scope)
      } else {
        if (!is.unit(value)) throwError('expect useUnit argument to be a unit')
        initial[key] = scope ? scopeBind(value as Event<any>, {scope}) : value
      }
    }
    return [{ref: initial}, storeMap, stores]
  }, [...entries.flat(), scope])
  const subscribe = React.useCallback(
    (cb: () => void) => {
      const seq = [
        step.compute({
          fn(value, _, stack) {
            const storeId = stack.parent!.node.id
            const storeKey = storeMap[storeId]
            current.ref = isList
              ? [...(current.ref as any[])]
              : {...current.ref}
            current.ref[storeKey] = value
          },
        }),
        /**
         * 'effect' priority cannot be batched
         * so we wait other effects to finish and then run batching
         * this is not an essential to work
         * just a fine tuning of a really low priority (ui) update
         *
         * also note that step.run({fn}) is an alias for
         * step.compute({fn, priority: 'effect'})
         * and since 22 version fn might be ommited if not used
         * */
        step.compute({priority: 'effect'}),
        step.compute({priority: 'sampler', batch: true}),
        step.run({fn: () => cb()}),
      ]
      if (scope) {
        const node = createNode({node: seq})
        const scopeLinks: {[_: string]: Node[]} = (scope as any).additionalLinks
        const storeLinks: Node[][] = []
        stores.forEach(store => {
          const id = (store as any).graphite.id
          const links = scopeLinks[id] || []
          scopeLinks[id] = links
          links.push(node)
          storeLinks.push(links)
        })
        return () => {
          storeLinks.forEach(links => {
            const idx = links.indexOf(node)
            if (idx !== -1) links.splice(idx, 1)
          })
          clearNode(node)
        }
      } else {
        const node = createNode({
          node: seq,
          parent: stores,
          family: {owners: stores},
        })
        return () => {
          clearNode(node)
        }
      }
    },
    [current],
  )
  const read = React.useCallback(
    () => (isShape ? current.ref : current.ref.unit),
    [current],
  )
  return useSyncExternalStore(subscribe, read, read)
}

export function useStoreMapBase<State, Result, Keys extends ReadonlyArray<any>>(
  [configOrStore, separateFn]: [
    configOrStore:
      | {
          store: Store<State>
          keys: Keys
          fn(state: State, keys: Keys): Result
          updateFilter?: (update: Result, current: Result) => boolean
        }
      | Store<State>,
    separateFn?: (state: State, keys: Keys) => Result,
  ],
  scope?: Scope,
): Result {
  let fn: (state: State, keys: Keys) => Result
  let updateFilter: (update: Result, current: Result) => boolean =
    basicUpdateFilter
  let store: Store<State>
  let keys: Keys
  if (separateFn) {
    fn = separateFn
    store = configOrStore as Store<State>
    keys = [] as unknown as Keys
  } else {
    fn = (configOrStore as any).fn
    store = (configOrStore as any).store
    keys = (configOrStore as any).keys
    updateFilter = (configOrStore as any).updateFilter || basicUpdateFilter
  }
  if (!is.store(store)) throwError('useStoreMap expects a store')
  if (!Array.isArray(keys)) throwError('useStoreMap expects an array as keys')
  if (typeof fn !== 'function') throwError('useStoreMap expects a function')

  const subscribe = React.useCallback(
    (cb: () => void) => createWatch(store, cb, scope),
    [store, scope],
  )
  const read = React.useCallback(
    () => stateReader(store, scope),
    [store, scope],
  )

  const stateRef = React.useRef<State>()
  const valueRef = React.useRef<Result>()
  const keysRef = React.useRef(keys)

  const value = useSyncExternalStoreWithSelector(
    subscribe,
    read,
    read,
    state => {
      if (stateRef.current !== state || !keysEqual(keysRef.current, keys)) {
        const result = fn(state, keys)

        stateRef.current = state
        keysRef.current = keys

        // skip update, if undefined
        // just like original store or previous implementation
        if (result !== undefined) {
          valueRef.current = result
        }
      }

      return valueRef.current as Result
    },
    (current, update) => !updateFilter(update, current),
  )

  return value
}
export function useListBase<T>(
  list: Store<T[]>,
  renderItem:
    | {
        keys?: any[]
        fn(item: T, index: number): React.ReactNode
        getKey?: (item: T) => string
      }
    | ((item: T, index: number) => React.ReactNode),
  scope?: Scope,
): React.ReactNode {
  let keys = [] as any[]
  let fn
  let getKey: (item: T) => string
  if (typeof renderItem === 'object' && renderItem !== null) {
    if (renderItem.keys) keys = renderItem.keys
    fn = renderItem.fn
    if (renderItem.getKey) getKey = renderItem.getKey
  } else {
    fn = renderItem
  }
  if (!is.store(list)) throwError('expect useList first argument to be a store')
  if (typeof fn !== 'function')
    throwError("expect useList's renderItem to be a function")
  if (!Array.isArray(keys)) throwError("expect useList's keys to be an array")
  const Item = React.useMemo(() => {
    const Item = withDisplayName(
      `${list.shortName || 'Unknown'}.Item`,
      (
        props:
          | {index: number; keys: any[]; keyVal: never; value: never}
          | {index: never; keys: any[]; keyVal: string; value: T},
      ) => {
        const {index, keys, keyVal, value} = props
        const isKeyed = !!fnRef.current[1]
        if (isKeyed) {
          return fnRef.current[0](value, keyVal as any)
        }
        const item = useStoreMapBase(
          [
            {
              store: list,
              keys: [index, ...keys],
              fn: (list, keys) => list[keys[0]],
            },
          ],
          scope,
        )
        return fnRef.current[0](item, index)
      },
    )
    return React.memo(Item)
  }, [list, scope, !!getKey!])
  const fnRef = React.useRef([fn, getKey!] as const)
  fnRef.current = [fn, getKey!]
  const keysSelfMemo = React.useMemo(() => keys, keys)
  if (getKey!) {
    return useStoreBase(list, scope).map(value => {
      const key = fnRef.current[1](value)
      return React.createElement(Item, {
        keyVal: key,
        key,
        keys: keysSelfMemo,
        value,
      })
    })
  } else {
    const length = useStoreMapBase(
      [
        {
          store: list,
          keys: [list],
          fn: list => list.length,
        },
      ],
      scope,
    )
    return Array.from({length}, (_, i) =>
      React.createElement(Item, {
        index: i,
        key: i,
        keys: keysSelfMemo,
      }),
    )
  }
}
