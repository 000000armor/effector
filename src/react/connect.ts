import React from 'react'

import {Store} from 'effector'
import {createReactState} from './createReactState'

export const connect = <State>(Component: React.ComponentType<any>) => (
  store: Store<State>,
) => createReactState(store, Component)
