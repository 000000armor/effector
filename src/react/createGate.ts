import React from 'react'
import {createStore, launch, Store, Domain, createEvent} from 'effector'
import {Gate} from './index.h'
import {useIsomorphicLayoutEffect} from './useIsomorphicLayoutEffect'
import {withDisplayName} from './withDisplayName'
import {flattenConfig, processArgsToConfig} from '../effector/config'
import {isObject} from '../effector/is'

export function useGate<Props>(
  GateComponent: Gate<Props>,
  props: Props = {} as any,
) {
  const propsRef = React.useRef<{value: any; count: number}>({
    value: null,
    count: 0,
  })
  useIsomorphicLayoutEffect(() => {
    GateComponent.open(propsRef.current.value)
    return () => GateComponent.close(propsRef.current.value) as any
  }, [GateComponent])
  if (!shallowCompare(propsRef.current.value, props)) {
    propsRef.current.value = props
    propsRef.current.count += 1
  }
  useIsomorphicLayoutEffect(() => {
    GateComponent.set(propsRef.current.value)
  }, [propsRef.current.count])
}

function shallowCompare(a: any, b: any) {
  if (a === b) return true
  if (
    typeof a === 'object' &&
    a !== null &&
    typeof b === 'object' &&
    b !== null
  ) {
    const aKeys = Object.keys(a)
    const bKeys = Object.keys(b)
    if (aKeys.length !== bKeys.length) return false
    for (let i = 0; i < aKeys.length; i++) {
      const key = aKeys[i]
      if (a[key] !== b[key]) return false
    }
    return true
  }
  return false
}
export function createGateImplementation<State>({
  domain,
  defaultState,
  hook: useGateHook,
  mainConfig,
  maybeConfig,
}: {
  domain?: Domain
  defaultState: State | {}
  hook: typeof useGate
  mainConfig?: Record<string, any>
  maybeConfig?: Record<string, any> & {sid?: string}
}): Gate<State> {
  const config = flattenConfig({
    or: maybeConfig,
    and: mainConfig,
  }) as {sid: string | undefined; name: string | undefined}
  const name = config.name || 'gate'
  const fullName = `${domain ? `${domain.compositeName.fullName}/` : ''}${name}`
  const set = createEvent<State>({
    name: `${fullName}.set`,
    sid: config.sid ? `${config.sid}|set` : undefined,
  })
  const open = createEvent<State>({
    name: `${fullName}.open`,
    sid: config.sid ? `${config.sid}|open` : undefined,
  })
  const close = createEvent<State>({
    name: `${fullName}.close`,
    sid: config.sid ? `${config.sid}|close` : undefined,
  })
  const status = createStore(Boolean(false), {
    name: `${fullName}.status`,
    // doesn't need to have sid, because it is internal store, should not be serialized
  })
    .on(open, () => Boolean(true))
    .on(close, () => Boolean(false))
  const state = createStore(defaultState as State, {
    name: `${fullName}.state`,
    sid: config.sid,
  })
    .on(set, (_, state) => state)
    .on(open, (_, state) => state)
    .reset(close)
  if (domain) {
    const {hooks} = domain as any
    launch({
      target: [
        hooks.store,
        hooks.store,
        hooks.event,
        hooks.event,
        hooks.event,
      ] as any,
      params: [status, state, open, close, set],
    })
  }
  function GateComponent(props: State) {
    useGateHook(GateComponent as any, props)
    return null
  }
  GateComponent.open = open
  GateComponent.close = close
  GateComponent.status = status
  GateComponent.state = state
  GateComponent.set = set
  return withDisplayName(`Gate:${fullName}`, GateComponent)
}

export function isPluginConfig(config: Record<string, any> | string) {
  return typeof config === 'object' && config !== null && 'sid' in config
}
export function isGateConfig(config: Record<string, any> | string) {
  return (
    typeof config === 'object' &&
    config !== null &&
    ('domain' in config || 'defaultState' in config || 'name' in config)
  )
}

export function createGate<Props>(...args: unknown[]): Gate<Props> {
  const [[nameOrConfig, defaultStateOrConfig], metadata] =
    processArgsToConfig(args)

  let domain
  let defaultState = {}
  let mainConfig = {}
  let maybeConfig = metadata

  if (typeof nameOrConfig === 'string') {
    mainConfig = {name: nameOrConfig}
    if (isPluginConfig(defaultStateOrConfig)) {
      // maybeConfig = defaultStateOrConfig
    } else {
      defaultState = defaultStateOrConfig || {}
    }
  } else if (isGateConfig(nameOrConfig)) {
    mainConfig = nameOrConfig
    defaultState = nameOrConfig.defaultState || {}
    domain = nameOrConfig.domain
  }

  return createGateImplementation<Props>({
    hook: useGate,
    domain,
    defaultState,
    mainConfig,
    maybeConfig,
  })
}
