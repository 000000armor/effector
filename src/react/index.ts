export {createStoreConsumer} from './createStoreConsumer'
export {createComponent} from './createComponent'
export {connect} from './connect'
export {createGate, useGate} from './createGate'
export {useEvent, useList, useStore, useStoreMap, useUnit} from './nossr'
