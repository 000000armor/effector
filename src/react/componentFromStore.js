//@flow

import * as React from 'react'
import {
  createEvent,
  createStore,
  type Event,
  type Store,
  type Subscription,
} from 'effector'

export function componentFromStore<Props: {...}>(
  propsToVdom: (props: Store<Props>) => Store<React.Node>,
) {
  class ComponentFromStore extends React.Component<Props, {vdom: React.Node, ...}> {
    state /*: {vdom: React.Node, ...}*/ = {vdom: null}

    propsEmitter: Event<Props> = createEvent()

    // Stream of props
    props$: Store<Props> = createStore((null: any))
      .on(this.propsEmitter, (_, props) => props)
      .map(props => {
        if (props) return props
        /*::return props*/
      })

    // Stream of vdom
    vdom$ = propsToVdom(this.props$)

    subscription: Subscription

    componentWillMount() {
      // Subscribe to child prop changes so we know when to re-render
      this.subscription = this.vdom$.watch(vdom => {
        this.setState({vdom})
      })
      this.propsEmitter(this.props)
    }

    componentWillReceiveProps(nextProps: Props) {
      // Receive new props from the owner
      this.propsEmitter(nextProps)
    }

    shouldComponentUpdate(
      nextProps: Props,
      nextState /*: {vdom: React.Node, ...}*/,
    ) {
      return nextState.vdom !== this.state.vdom
    }

    componentWillUnmount() {
      this.props$.off(this.propsEmitter)
      this.subscription.unsubscribe()
    }

    render() {
      return this.state.vdom
    }
  }
  return ComponentFromStore
}
