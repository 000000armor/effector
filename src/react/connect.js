//@flow

import * as React from 'react'

import type {Store} from 'effector'
import {createReactState} from './createReactState'

export function connect<State: Object, Com: React.ComponentType<*>>(
  Component: Com,
): (
  store: Store<State>,
) => React.ComponentType<$Exact<$Diff<React.ElementConfig<Com>, State>>> {
  return store => createReactState(store, Component)
}
