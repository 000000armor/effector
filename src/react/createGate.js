//@flow

import * as React from 'react'
import {createStore, createApi, launch, type Store, type Event} from 'effector'
import {useIsomorphicLayoutEffect} from './useIsomorphicLayoutEffect'
import {withDisplayName} from './withDisplayName'

export type Gate<Props = {||}> = Class<React.Component<Props>> &
  interface {
    isOpen: boolean,
    isTerminated: boolean,
    open: Event<Props>,
    close: Event<Props>,
    status: Store<boolean>,
    destructor: Event<void>,
    current: Props,
    state: Store<Props>,
    set: Event<Props>,
    childGate<Next>(childName?: string): Gate<Next>,
  }

export function useGate<Props>(
  GateComponent: Gate<Props>,
  props?: Props = ({}: any),
) {
  const propsRef = React.useRef(null)
  useIsomorphicLayoutEffect(() => {
    GateComponent.open(propsRef.current)
    return () => GateComponent.close(propsRef.current)
  }, [GateComponent])
  if (!shallowCompare(propsRef.current, props)) {
    propsRef.current = props
    GateComponent.set(props)
  }
}

function shallowCompare(a, b) {
  if (a === b) return true
  if (
    typeof a === 'object' &&
    a !== null &&
    typeof b === 'object' &&
    b !== null
  ) {
    const aKeys = Object.keys(a)
    const bKeys = Object.keys(b)
    if (aKeys.length !== bKeys.length) return false
    for (let i = 0; i < aKeys.length; i++) {
      const key = aKeys[i]
      if (a[key] !== b[key]) return false
    }
    return true
  }
  return false
}

export function createGate<Props>(
  name: string = 'gate',
  defaultState: Props = ({}: any),
): Gate<Props> {
  let domain
  if (typeof name === 'object' && name !== null) {
    if ('defaultState' in name) {
      defaultState = name.defaultState
    }
    if (name.domain) domain = name.domain
    name = name.name || 'gate'
  }
  const status = createStore(Boolean(false))
  const state: Store<Props> = createStore(defaultState)
  const {set} = createApi(state, {
    set: (_, state) => state,
  })

  const {open, close, destructor} = createApi(status, {
    open: () => GateComponent.predicate() && Boolean(true),
    close: () => Boolean(false),
    destructor: () => Boolean(false),
  })
  function GateComponent(props) {
    useGate(GateComponent, props)
    return null
  }
  GateComponent.predicate = () => Boolean(true)
  GateComponent.isOpen = Boolean(false)
  GateComponent.current = state.getState()
  GateComponent.open = open
  GateComponent.close = close
  GateComponent.status = status
  GateComponent.state = state
  GateComponent.set = set
  GateComponent.destructor = destructor
  GateComponent.isTerminated = Boolean(false)
  GateComponent.childGate = (childName: string = 'Subgate') => {
    console.error('childGate is deprecated')
    const gate = createGate(`${name}/${childName}`)
    ;(gate: any).predicate = () => GateComponent.status.getState()
    let isOpen = false
    let isFeedback = false
    gate.open.watch(() => {
      if (!isFeedback) isOpen = true
    })
    gate.close.watch(() => {
      if (!isFeedback) isOpen = false
    })

    GateComponent.status.watch(status => {
      isFeedback = true
      if (isOpen && status && !gate.status.getState()) {
        gate.open()
      }
      isFeedback = false
    })
    GateComponent.close.watch(() => {
      isFeedback = true
      gate.close()
      isFeedback = false
    })
    GateComponent.destructor.watch(() => gate.destructor())
    return gate
  }
  const unwatch = status.watch(status => (GateComponent.isOpen = status))
  const unwatch2 = state.watch(current => (GateComponent.current = current))
  const unwatch3 = status.map(status => {
    if (!status) GateComponent.current = defaultState
    return null
  })
  state.reset(close)

  const setTerminated = destructor.watch(() => {
    GateComponent.isTerminated = Boolean(true)
  })
  const unstate = () => {
    GateComponent.status.off(GateComponent.open)
    GateComponent.status.off(GateComponent.close)
    GateComponent.status.off(GateComponent.destructor)
  }

  const undestruct = destructor.watch(() => {
    undestruct()
    unwatch()
    unwatch2()
    setTerminated()
    unstate()
    GateComponent.state.off(GateComponent.set)
  })
  if (domain) {
    const {hooks} = domain
    launch({
      target: [hooks.store, hooks.store, hooks.event, hooks.event, hooks.event],
      params: [status, state, open, close, set],
    })
  }
  return withDisplayName(`Gate:${name}`, GateComponent)
}
