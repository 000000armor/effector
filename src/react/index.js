//@flow

import * as React from 'react'

import type {Store} from 'effector'

export function createReactState<State>(
 store: Store<State>,
): {
 Provider: React.ComponentType<{children: React.Node}>,
 Consumer: React.ComponentType<{
  children: (current: {
   value: State,
   ref: Store<State>,
  }) => React.Node,
 }>,
} {
 const {Provider, Consumer} = React.createContext({
  value: store.getState(),
  ref: store,
 })

 class StoreProvider extends React.Component<
  {
   children: React.Node,
  },
  {
   current: {
    value: State,
    ref: Store<State>,
   },
  },
 > {
  state = {
   current: {
    value: store.getState(),
    ref: store,
   },
  }
  componentDidMount() {
   store.watch(value => this.setState({current: {value, ref: store}}))
  }
  render() {
   return <Provider value={this.state.current}>{this.props.children}</Provider>
  }
 }

 return {
  Provider: StoreProvider,
  Consumer,
 }
}

export function createStoreComponent<
 State: Object,
 Com: React.ComponentType<*>,
>(
 store: Store<State>,
 Component: Com,
): React.ComponentType<$Exact<$Diff<React.ElementConfig<Com>, State>>> {
 const {Provider, Consumer} = createReactState(store)
 const ConnectedComponent = props => (
  <Provider>
   <Consumer>{({value}) => <Component {...props} {...value} />}</Consumer>
  </Provider>
 )
 ConnectedComponent.displayName = `Connect(${Component.displayName
  || 'Unknown'})`
 return ConnectedComponent
}

export function connect<State: Object, Com: React.ComponentType<*>>(
 Component: Com,
): (
 store: Store<State>,
) => React.ComponentType<$Exact<$Diff<React.ElementConfig<Com>, State>>> {
 return store => createStoreComponent(store, Component)
}
