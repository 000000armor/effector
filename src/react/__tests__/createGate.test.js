//@flow

import {configure} from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

configure({
  adapter: new Adapter(),
})

import * as React from 'react'
import {render, cleanup, container} from 'effector/fixtures/react'
import {mount} from 'enzyme'
import {createGate, useGate} from '../createGate'

test('plain gate', () => {
  const Gate = createGate('plain gate')
  expect(Gate.isOpen).toBe(false)
  const tree = mount(
    <section>
      <div>div</div>
      <Gate />
    </section>,
  )
  expect(Gate.isOpen).toBe(true)
  tree.unmount()
  expect(Gate.isOpen).toBe(false)
})

test('plain gate hook', async () => {
  const Gate = createGate('plain gate')
  expect(Gate.isOpen).toBe(false)
  const Component = () => {
    useGate(Gate)
    return (
      <section>
        <div>div</div>
      </section>
    )
  }
  await render(<Component />)

  expect(Gate.isOpen).toBe(true)
  await cleanup()
  expect(Gate.isOpen).toBe(false)
})

test('gate with props', async () => {
  const Gate = createGate('gate with props')
  expect(Gate.current).toMatchObject({})
  const tree = mount(
    <section>
      <Gate foo="bar" />
    </section>,
  )
  tree.render()
  expect(Gate.state.getState()).toMatchObject({foo: 'bar'})
  expect(Gate.current).toMatchObject({foo: 'bar'})
  expect(tree.text()).toMatchInlineSnapshot(`""`)
  tree.unmount()
  expect(Gate.state.getState()).toMatchObject({})
})

test('gate with props hook', async () => {
  const Gate = createGate('gate with props')
  expect(Gate.current).toMatchObject({})
  const Component = () => {
    useGate(Gate, {foo: 'bar'})
    return <section />
  }
  await render(<Component />)
  expect(Gate.state.getState()).toMatchObject({foo: 'bar'})
  expect(Gate.current).toMatchObject({foo: 'bar'})
  expect(container.firstChild).toMatchInlineSnapshot(`<section />`)
  await cleanup()
  expect(Gate.state.getState()).toMatchObject({})
})

function calls(fn, ...args) {
  expect(fn.mock.calls).toEqual(args.map(a => [a]))
}
test('gate properties', async () => {
  const Gate = createGate('gate properties')
  const fn1 = jest.fn()
  const fn2 = jest.fn()
  Gate.status.watch(isOpen => fn1(isOpen))
  Gate.state.watch(props => fn2(props))
  const tree = mount(
    <section>
      <Gate foo="bar" />
    </section>,
  )
  tree.render()
  tree.unmount()
  calls(fn1, false, true, false)
  calls(fn2, {}, {foo: 'bar'}, {})
})

test('gate properties hook', async () => {
  const Gate = createGate('gate properties')
  const fn1 = jest.fn()
  const fn2 = jest.fn()
  Gate.status.watch(isOpen => fn1(isOpen))
  Gate.state.watch(props => fn2(props))
  const Component = () => {
    useGate(Gate, {foo: 'bar'})
    return <section />
  }
  await render(<Component />)
  await cleanup()
  calls(fn1, false, true, false)
  calls(fn2, {}, {foo: 'bar'}, {})
})

describe('child gate', () => {
  test('usage', async () => {
    const Gate = createGate('parent gate')
    const Child = Gate.childGate('child gate')

    const tree = mount(
      <section>
        <Gate />
        <div>
          <Child />
        </div>
      </section>,
    )
    expect(Gate.isOpen).toBe(true)
    expect(Child.isOpen).toBe(true)
    expect(tree.text()).toMatchInlineSnapshot(`""`)
    tree.unmount()
    expect(Gate.isOpen).toBe(false)
    expect(Child.isOpen).toBe(false)
  })
  test('order edge case', async () => {
    const Gate = createGate('parent gate')
    const Child = Gate.childGate('child gate')

    const tree = mount(
      <section>
        <div>
          <Child />
        </div>
        <Gate />
      </section>,
    )
    expect(Gate.isOpen).toBe(true)
    expect(Child.isOpen).toBe(true)
    tree.unmount()
    expect(Gate.isOpen).toBe(false)
    expect(Child.isOpen).toBe(false)
  })
  test('parent prevent children from beeing open', async () => {
    const Gate = createGate('parent gate')
    const Child = Gate.childGate('child gate')

    await render(
      <section>
        <Child />
      </section>,
    )

    expect(Child.isOpen).toBe(false)
    await cleanup()
    expect(Child.isOpen).toBe(false)
  })
})
