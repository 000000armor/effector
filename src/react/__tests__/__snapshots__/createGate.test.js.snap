// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`child gate usage: child gate usage 1`] = `""`;

exports[`gate with props: gate with props 1`] = `""`;
